source/ui/memory-component-presenter.cpp

```diff
@@ -100,58 +121,126 @@ MemoryComponentPresenter::data(const QModelIndex &index, int role) const {
   // check boundaries
   assert::that(index.isValid());
 
+  switch (role) {
+    case AddressRole: return dataAdress(index, role);
+    case InfoRole: return dataInfo(index, role);
+    default: return dataMemory(index, role);
+  }
+}
+
+
+QVariant
+MemoryComponentPresenter::dataAdress(const QModelIndex &index, int role) const {
+  // format address as hex value and return it
+  return QString("%1")
+      .arg(index.row(), 4, 16, QLatin1Char('0'))
+      .toUpper()
+      .prepend("0x");
+}
+
+
+QVariant
+MemoryComponentPresenter::dataMemory(const QModelIndex &index, int role) const {
   // get role as a string because there is more information in it
   QString role_string = MemoryComponentPresenter::roleNames().value(role);
 
   int number_of_bits = 8;
   if (role_string.endsWith("16")) number_of_bits = 16;
   if (role_string.endsWith("32")) number_of_bits = 32;
 
-  if (role_string.startsWith("address")) {
-    // format index as hex value and return it
-    return QString("%1")
-        .arg(index.row() * number_of_bits / 8, 4, 16, QLatin1Char('0'))
-        .toUpper()
-        .prepend("0x");
-  }
 
-  if (role == InfoRole) {
-    // TODO fetch value from core
+  int memory_address = index.row();//* (number_of_bits / 8);
+  int memory_length = number_of_bits / 8;
+
+
+  if (memory_address + memory_length <= _memorySize) {
+    MemoryValue memory_cell =
+        getMemoryValueCached(memory_address, memory_length);
+
+    std::string memoryStringValue;
+    if (role_string.startsWith("bin")) {
+      memoryStringValue = StringConversions::toBinString(memory_cell);
+    }
+    // not yet implemented by conversions
+    // else if (role_string.startsWith("oct"))
```

Comment:
I would like to keep the octal conversions commented out. It thinks it's quite likely that someone could implement this feature in the future.